op-supervisor/supervisor/backend/db/fromda/update.go

```diff
@@ -226,15 +230,17 @@ func (db *DB) addLink(source eth.BlockRef, derived eth.BlockRef, invalidated com
 	if lastDerived.Number == derived.Number {
 		// Same block height? Then it must be the same block.
 		// I.e. we encountered an empty L1 block, and the same L2 block continues to be the last block that was derived from it.
-		if invalidated != (common.Hash{}) {
-			if lastDerived.Hash != invalidated {
-				return fmt.Errorf("inserting block %s that invalidates %s at height %d, but expected %s: %w", derived.Hash, invalidated, lastDerived.Number, lastDerived.Hash, types.ErrConflict)
+		if invalidRef != (common.Hash{}) {
+			if lastDerived.Hash != invalidRef {
+				return fmt.Errorf("inserting block %s that invalidates %s at height %d, but expected %s: %w", derived.Hash, invalidRef, lastDerived.Number, lastDerived.Hash, types.ErrConflict)
 			}
 		} else {
 			if lastDerived.Hash != derived.Hash {
 				return fmt.Errorf("derived block %s conflicts with known derived block %s at same height: %w",
 					derived, lastDerived, types.ErrConflict)
 			}
+			// invalidRef is used and references pass, so this is actually a replacement block
+			link.entryType = ReplacementV0
```

Comment:
This comment makes no sense to me, can you explain it?
This happens in the `else` of the `invalidRef != (common.Hash{})`, so `invalidRef` is *not* being used here.

And we just checked that we're deriving the same L2 block-hash as last known, so the L2 block is not being replaced, but the L1 block source is changing.

Is this `link.entryType = ReplacementV0` meant to go in a different place?

Reply:
Ah, my bad, this was meant to be located ~5 lines higher, when the ref is confirmed to match the `lastDerived`. Not sure why I ended up putting it lower.

In general I've had some issues parsing the full logic of `addLink` since it's quite long and has a lot of branches. Anyway, my mistake, will fix.

Reply:
I am going over the function more and I'm thinking I'll move this check to the top of the function, next to the other type settings. I figure we know it's a replacement if the `invalid` is set but doesn't match the `derived` provided.

I think there's a function comment somewhere that was misleading, about how `invalid` must always match derived. If I see that again I'll try to clear it up.

Reply:
Ya, I made this a much nice type switch at the top of the function, and got out of the way of the consistency checks. Much cleaner IMO. And made the comment a little more clear about the different modalities with `invalidRef`